### PR TITLE
Added math equation support for plus and minus signs.

### PR DESCRIPTION
**Pull request by mathieu-fournier**
Pull request for the issue feature #6935 

## Added initial support for parsing math equations. + and - 
Plus and minus signs are supported.
Ex: 
5 ft + 12 in + 1 yd -1 ft 6 in = 7.5 ft
(5 + 1 + 3 - 1.5)
Multiplication and division are not supported yet. 

## Special case + Refactoring
I noticed that the functions `createQuantityFromParseTokens` and `getQuantityValueFromParseTokens` where doing the same "tokens to value" logic. So I de-duplicated them.  
There was one inconsistent output between these methods. 
The case where only the unit label was specified. <- See line 601 in `Parser.ts`.
The output for the 2 methods where :   'FT' = 1   and    'FT' = 0
I opted to assume that the magnitude is 1. 

Another special case, which I assume was a bug, is the parsing of `2'-6"`. <- See line 218 in the file `Parsing.test.ts`. 
The minus sign was threated like a formatTraits `fractionDash`, but the format did not specified `fractionDash`, so it should've been 0 or error. 
Now it's treated like a minus sign when fractionDash is NOT specified, and a fractionDash when fractionDah is specified.

### CODE REVIEW

File: core/quantity/src/Parser.ts
Comment (hl662) [inline core/quantity/src/Parser.ts:50]:
```suggestion
  subtraction = "-",
```

**Comment by anmolshres98** [comment]
> ## Added initial support for parsing math equations. + and -
> Plus and minus signs are supported. Ex: 5 ft + 12 in + 1 yd -1 ft 6 in = 7.5 ft (5 + 1 + 3 - 1.5) Multiplication and division are not supported yet.

In cases like these where multiple metrics are supplied (yd, ft, in, m, cm, ...) how do you decide on which to go with?

File: core/quantity/src/Parser.ts
Comment (hl662) [inline core/quantity/src/Parser.ts:512]: Can you add a test case that explicitly demonstrates this behavior?

Comment (mathieu-fournier) [reply core/quantity/src/Parser.ts:512]: I added one, thanks

**Review by hl662 — APPROVED** [review]

**Review by aruniverse — CHANGES_REQUESTED** [review]
Want to hold this PR and discuss motivations before merging this in

File: core/quantity/src/test/Parsing.test.ts
Comment (ColinKerr) [inline core/quantity/src/test/Parsing.test.ts:213]: why did you have to change this case?

Comment (aruniverse) [reply core/quantity/src/test/Parsing.test.ts:213]: add test case for same exact string where `-` is a `spacer`

File: core/quantity/src/test/Parsing.test.ts
Comment (aruniverse) [inline core/quantity/src/test/Parsing.test.ts:403]: revert this back to 0 magnitude

Comment (mathieu-fournier) [reply core/quantity/src/test/Parsing.test.ts]: Done, thx

File: core/quantity/src/test/Parsing.test.ts
Comment (mathieu-fournier) [reply core/quantity/src/test/Parsing.test.ts:213]: Done, thx

File: core/quantity/src/Formatter/Interfaces.ts
Comment (hl662) [inline core/quantity/src/Formatter/Interfaces.ts:28]: Can you add a snippet to NextVersion.md detailing how to enable mathematic equations? What operations are supported, what happens if you pass in an unsupported operator like multiplication?

Comment (mathieu-fournier) [reply core/quantity/src/Formatter/Interfaces.ts:28]: Sure, I'll do that (:

Comment (hl662) [reply core/quantity/src/Formatter/Interfaces.ts:28]: Adding to this, can a test case be added for multiplication? Should a case like "1 FT * 6IN" throw ParseError.MathematicEquationFoundButIsNotAllowed?

Comment (mathieu-fournier) [reply core/quantity/src/Formatter/Interfaces.ts:28]: I cannot add a test for that right now since these operators are not "tokenized". 
Adding "tokenization" for these operators is half the effort for a full support, it's considerable. 

With the current implementation, it will result in an error anyway, just not this one : ParseError.MathematicEquationFoundButIsNotAllowed.

Comment (mathieu-fournier) [reply core/quantity/src/Formatter/Interfaces.ts:28]: Done with the summary, let me know if I'm missing anything.

**Review by anmolshres98 — CHANGES_REQUESTED** [review]

File: core/quantity/src/Formatter/Format.ts
Comment (mathieu-fournier) [reply core/quantity/src/Formatter/Format.ts:39]: It only support addition and subtractions so it would be misleading to tell otherwise here.

File: docs/changehistory/NextVersion.md
Comment (mathieu-fournier) [reply docs/changehistory/NextVersion.md:65]: It doesnt throw, it returns a ParsingError in one case, in the other is returns an invalid Quantity.

Comment (anmolshres98) [reply docs/changehistory/NextVersion.md:65]: aah ok, then could you write "Other operators will end up returning" then? thanks

File: core/quantity/src/Formatter/Format.ts
Comment (anmolshres98) [reply core/quantity/src/Formatter/Format.ts:39]: ok the please change to:
```ts
protected _allowMathematicEquations: boolean = false; // optional; enables doing maths like addition and subtraction; default is false.
```

File: docs/changehistory/NextVersion.md
Comment (mathieu-fournier) [reply docs/changehistory/NextVersion.md:65]: yea sure

**Review by anmolshres98 — APPROVED** [review]

File: core/quantity/src/Formatter/Format.ts
Comment (aruniverse) [inline core/quantity/src/Formatter/Format.ts:39]:
```suggestion
  protected _allowMathematicEquations: boolean = false; // optional; enables calculating mathematic operations like addition and subtraction; default is false.
```

File: common/changes/@itwin/core-quantity/pr-mathieu-fournier-6969_2024-07-29-18-30.json
Comment (aruniverse) [inline common/changes/@itwin/core-quantity/pr-mathieu-fournier-6969_2024-07-29-18-30.json:5]:
```suggestion
      "comment": "Add optional property `allowMathematicEquations` to `BaseFormat` to calculate mathematical operations like addition and subtraction.",
```

File: docs/changehistory/NextVersion.md
Comment (aruniverse) [inline docs/changehistory/NextVersion.md:55]: This is really parsing mathematical operations and calculating it, not formatting right?

```suggestion
## Mathematical operation parsing
```

File: docs/changehistory/NextVersion.md
Comment (aruniverse) [inline docs/changehistory/NextVersion.md:57]:
```suggestion
The quantity formatter now supports parsing mathematical operations. Ex :
```

File: docs/changehistory/NextVersion.md
Comment (aruniverse) [inline docs/changehistory/NextVersion.md:58]: where is the actual ts code for this? what does it look like?

File: docs/changehistory/NextVersion.md
Comment (aruniverse) [inline docs/changehistory/NextVersion.md:63]:
```suggestion
### Limitations
```

File: docs/changehistory/NextVersion.md
Comment (aruniverse) [inline docs/changehistory/NextVersion.md:64]:
```suggestion
Only plus(`+`) and minus(`-`) signs are supported for now.
```

**Review by aruniverse — COMMENTED** [review]
nit, maybe this is just an issue of wording or mu interpretation,

should it be mathematic **operations** not **equations** ? if so all references to the boolean flag you added need to be updated

File: docs/changehistory/NextVersion.md
Comment (mathieu-fournier) [reply docs/changehistory/NextVersion.md:55]: You're right, it should be operation everywhere, equations are algebra stuff. Thanks !

File: docs/changehistory/NextVersion.md
Comment (mathieu-fournier) [reply docs/changehistory/NextVersion.md:58]: it's in the tests. 
It looks like this : 
```
// Async version 
const quantityProps = await Parser.parseIntoQuantity("2 FT 6 in + 9", format, unitsProvider);

// Sync version
const parseResult = Parser.parseToQuantityValue("2 FT 6 in + 9", formatU, meterConversionSpecs);
```

Comment (aruniverse) [reply docs/changehistory/NextVersion.md:58]: consumers will not be looking at the tests, theyll look at the the changelog here, and this should probably have a ts snippet of the change

Comment (mathieu-fournier) [reply docs/changehistory/NextVersion.md:58]: Users will use the quantity formatter with the boolean flag. 
They will not call these directly. Should I add an example anyway ?

**Comment by mathieu-fournier** [comment]
Moved to https://github.com/iTwin/itwinjs-core/pull/7019

**Pull request closed without merging**
